📖 Readme (#62)
overture-stack/iobio-components: packages/iobio-react-components/README.md

## Setup + Local Development

### Requirements:

- PNPM 9+
- Node 20+

> - Install dependencies: `pnpm i`

Reviewer: should this be `ci`? or does pnpm figure out all that npm `ci` vs `i` stuff itself

Author: PNPM is already managing that.

https://pnpm.io/cli/install
https://stackoverflow.com/questions/70154568/pnpm-equivalent-command-for-npm-ci

- Import components in consumer app as `import { IobioDataBroker } from '@overture-stack/iobio-components/packages/iobio-react-components/;`

### Build Package

- `pnpm run build`

## React Components

- This package exports a set of React + TypeScript components to facilitate the integration of the Iobio Charts library of "web components" into your UI app.

- The Demo App at `iobio-components/apps/bam.iobio` shows an example integration which closely mirrors the integration at bam.iobio.io

- The `<IobioDataBroker />` component is necessary to analyze a given file. It receives an `alignmentUrl` string prop which should point to a BAM/CRAM file URL. The Iobio server performs an analysis on the given file, and the other Iobio components read this data.

> 

Reviewer: I think `..and the other Iobio components read this data` could be expanded on. It's not immediately clear or a super common that the React components are mostly wrapping web components for display, which are hooked in to events in the data broker components. If I came across this repo, without seeing it being developed, I would assume React (or React compatible libraries) were controlling the state.

Author: OK added some clarified language

- Some visualization components (`histogram` and `percentBox`) require data keys that correspond to this analysis to know what data to read. These values can be found in `/utils/constants`

- The Overture `Iobio React Components` library also adds basic Typescript definitions in iobio.d.ts. `iobio-charts` is not typed.

### SSR Integrations

> - Integrations with Overture Stage revealed conflicts between web component usage in `iobio-charts` and Next.js server side rendering, which may be an issue in future projects. Iobio Charts uses DOM APIs that are not available when rendered on the server, and which create conflicts when rendered multiple times.

Reviewer 2: ```suggestion
- Integrations with Overture Stage revealed conflicts between web component usage in `iobio-charts` and Next.js server side rendering, which may be an issue in future projects. Iobio Charts uses a DOM API that is not available when rendered on the server, and which create conflicts when rendered multiple times.
```

> - These APIs include (but are not limited to `customElements`, `HTMLElement`, `Element` and `CssStyleSheet`.

Reviewer 2: ```suggestion
- This API is used by (but not limited to) `customElements`, `HTMLElement`, `Element` and `CssStyleSheet`.
```

- The Stage solution is to leverage `JSDOM` and add DOM shims to handle the server side code. `Iobio React Components` also exports an IIFE async import to enforce asynchronous imports.

- example: 

```
// next.config.js
module.exports = withPlugins([withTranspileModules], {
	webpack: (config, options) => {
		if (options.isServer) {
			const { JSDOM } = jsdom;
			const dom = new JSDOM('', { url: 'http://localhost/' });
      global.customElements = global.window.customElements;
```

> - This solution works in Next 12.1 and may be updated in future iterations.

Reviewer: I appreciate the detail in this and the further information for integration.
I do think it's a bit unfair to term these issues as "conflicts"

TLDR: let's just be explicit we only support client side rendering

Could we just be explicit and say these components only work client side? 
That's the extend of our work and what _we_ support with our lib.
Every framework has some form of SSR so I don't think these issues are unique to Next, we just don't support SSR.
And DOM apis are web browser not server apis.
This phrasing feels to me like we're pointing out bugs or issues. 
But the case is closer to "my gif won't play on this printed sheet of paper."

I think being **explicit that we only support client side rendering** should also cover integration for others with frameworks.

I don't feel great about recommending an entire DOM library dependency just for a frontend visualisation lib to work. It's also an extra dependency that isn't in the package.json of the package.

We're out of time now, but in the future maybe we could integrate it into our `iobio-component` lib because ,as I understand, right now it's a hard dependency to send `dom` event messages and "register" custom elements. It's also completely dependent on that library maintainer keeping up to speed with specs vs the browser companies. Take all of this with a grain of salt, I know we have limited time, just my 2cents.

Author: For context re: limited time, this is a break period which is ideal to identify issues for follow up when we resume dev work on this project.

Being explicit about 'client side only' is a very fair change to the readme.

This is included in the readme explicitly because it's an area for improvement and dev attention. I feel like there is a more optimal solution out there, but for use with Next 12 and our timeframes this worked. Ideally I'd update Stage to Next 14 to start.

I wanted to provide an example as a starting point. I don't think this is too out of line with other readmes I see on Github / NPM. I feel like guides for integration with different frameworks isn't unheard of.

It did cross my mind that the dom shim code in Stage next.config could possibly be shipped as an export from this library. So there's potential there to include 'jsdom' as a dependency and export the block from [Stage next.config](https://github.com/overture-stack/stage/blob/iobio/next.config.js#L19) as a helper function.

Sorry where are you referencing `DOM apis are web browser not server apis` ?

Author: I've rephrased this block but will review again.

> - This solution works in Next 12.1 and may be updated in future iterations.

Reviewer 3: Markdown formatting seems wrong here. There are comments outside the code block, and the code block will need triple back ticks.

Author: Fixed, thanks
